onnxruntime/core/providers/cuda/math/matmul.cc

```diff
@@ -88,6 +95,366 @@ static bool CanUseStridedBatchedGemm(const TensorShape& left_shape, const Tensor
   return true;
 }
 
+template <typename U>
+float ComputeStandardDeviation(const U* elems, const int32_t size)
+{
+  if (size == 0 || elems == nullptr) {
+    return 0.0f;
+  }
+
+  // Calculate the mean
+  float sum = 0.0f;
+  for (int32_t i = 0; i < size; i ++)
+    sum += float(elems[i]);
+  float mean = sum / size;
+
+  // Calculate the sum of squared differences from the mean
+  float squaredSum = 0.0f;
+  for (int32_t i = 0; i < size; i ++) {
+    float diff = float(elems[i]) - mean;
+    squaredSum += diff * diff;
+  }
+
+  // Calculate the variance
+  float variance = squaredSum / size;
+
+  // Return the square root of variance as standard deviation
+  return std::sqrt(variance);
+}
+
+Status ComputeScale(cudaStream_t& stream, const Tensor* tensor, const float std_quant, float& scale)
+{
+  const int32_t num_coef = tensor->Shape().Size();
+  MLFloat16* scale_coef = (MLFloat16*)malloc(num_coef * sizeof(MLFloat16));
+  auto status = ComputeStdDevCoefficientsForScale(stream, tensor, num_coef, scale_coef);
+  if (! status.IsOK())
+    return status;
+
+  float std_coef = ComputeStandardDeviation(scale_coef, num_coef);
+  free(scale_coef);
+
+  // If the standard deviation is 0, just use a scale of 1
+  scale = fabs(std_coef) < 1e-5 ? 1.0f : std_quant / std_coef;
+
+  return status;
+}
+
+void NoOpDeleter(void* [[maybe_unused]] ptr) {
+  (void)(ptr);
+}
+
+Status TransposeAndConvertToFp8(cudaStream_t& cuda_stream, const Tensor* right_X, IAllocatorUniquePtr<void>& right_X_fp8,
+  const cudaDeviceProp& device_prop, const AllocatorPtr& allocator, Stream* stream)
+{
+  const auto shape_size = right_X->Shape().GetDims().size();
+  const int num_elems = right_X->SizeInBytes() / sizeof(MLFloat16);
+
+  cublasHandle_t cublas_handle;
+  CUBLAS_RETURN_IF_ERROR(cublasCreate(&cublas_handle));
```

Comment:
I think you can pass cublas handle from context:
https://github.com/microsoft/onnxruntime/blob/0371e92419e4c6e93717c28aae2a3e8a36165bae/onnxruntime/contrib_ops/cuda/bert/multihead_attention.cc#L435

Otherwise, you will need set the compute stream for cublas handle like `cublasSetStream(handle, stream)`.

Reply:
The `TransposeAndConvertToFp8` is called from `PrePack`, and I don't seem to have access to the context there. So I will add the `cublasSetStream` call.

Reply:
updated